Add collection function docs (#1408)
neo4j/docs-cypher: modules/ROOT/pages/functions/list.adoc

  (bob)-[:MARRIED]->(eskil)
----

[[functions-coll-distinct]]
== coll.distinct()

.Details
|===
| *Syntax* 3+| `coll.distinct(list)`
| *Description* 3+| Returns the given list with all duplicate values removed.
.2+| *Arguments* | *Name* | *Type* | *Description*
| `list` | `LIST<ANY>` | A list to be deduplicated.
| *Returns* 3+| `LIST<ANY>`
|===

.Considerations
|===
| `coll.distinct(null)` returns `null`.

> | The order of the first occurrences is preserved.

Reviewer: ```suggestion
| The order of the first occurrences of the distinct values is preserved.
```

|===

.+coll.distinct()+
======

.Deduplicate a list of integer values

> // tag::functions_list_distinct_ints[]

Reviewer: ```suggestion
```

[source, cypher]
----
RETURN coll.distinct([1, 3, 2, 4, 2, 3, 1])
----

> // end::functions_list_distinct_ints[]

Reviewer: ```suggestion
```

A `LIST<ANY>` containing only unique values is returned.

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| coll.distinct([1, 3, 2, 4, 2, 3, 1])
| [1, 3, 2, 4]
1+d|Rows: 1
|===

.Deduplicate a list of mixed values, including null

> // tag::functions_list_distinct[]

Reviewer: this is the same tag as in line 61 - they should be different (or we only use one for cheat sheet inclusion)

edit: this also holds for other tags in this PR

> // tag::functions_list_distinct_mixed[]

Reviewer: ```suggestion
// tag::functions_list_distinct[]
```

[source, cypher]
----
RETURN coll.distinct([1, true, true, null, 'a', false, true, 1, null])
----

> // end::functions_list_distinct_mixed[]

Reviewer: ```suggestion
// end::functions_list_distinct[]
```

A `LIST<ANY>` containing only unique values is returned.

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| coll.distinct([1, true, true, null, 'a', false, true, 1, null])
| [1, true, null, 'a', false]
1+d|Rows: 1
|===

======

[[functions-coll-flatten]]
== coll.flatten()

.Details
|===
| *Syntax* 3+| `coll.flatten(list [, depth])`

> | *Description* 3+| Returns a list flattened to the given depth.

Reviewer: ```suggestion
| *Description* 3+| Returns a list flattened to the given nesting depth.
```

.3+| *Arguments* | *Name* | *Type* | *Description*

> | `list` | `LIST<ANY>` | A list to be flattened.

Reviewer: ```suggestion
| `list` | `LIST<ANY>` | A nested list to be flattened.
```

| `depth` | `INTEGER` | The maximum depth to flatten to (default value: 1).
| *Returns* 3+| `LIST<ANY>`
|===

.Considerations
|===
| Preserves the original order of non-list items.
| If `depth` is 0, the input list is returned unchanged.
| `depth` must be a non-negative integer.
| `coll.flatten(null)` returns `null`.
| `coll.flatten(list, null)` returns `null`.
| `coll.flatten(null, depth)` returns `null`.
| `coll.flatten(null, null)` returns `null`.

|===

.+coll.flatten()+
======

.Flatten a list containing nesting to level 2
// tag::functions_list_flatten[]
[source, cypher]
----
RETURN coll.flatten(['a', ['b', ['c']]], 2)
----
// end::functions_list_flatten[]

A `LIST<ANY>` with any inner lists up to a nesting depth of 2 flattened.

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| coll.flatten(['a', ['b', ['c']]], 2)
| ['a', 'b', 'c']
1+d|Rows: 1
|===

.Flatten a list to default of depth 1

> // tag::functions_list_flatten_default[]

Reviewer: ```suggestion
```

[source, cypher]
----
RETURN coll.flatten(['a', ['b', ['c']]])
----

> // end::functions_list_flatten_default[]

Reviewer: ```suggestion
```

A `LIST<ANY>` with any inner lists up to a nesting depth of 1 (default depth) flattened.

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| coll.flatten(['a', ['b', ['c']]])
| ['a', 'b', ['c']]
1+d|Rows: 1
|===
======

[[functions-coll-indexOf]]
== coll.indexOf()

.Details
|===
| *Syntax* 3+| `coll.indexOf(list, value)`

> | *Description* 3+| Returns the index for the first match of value in the given list, if the value is not present, -1 is returned.

Reviewer: ```suggestion
| *Description* 3+| Returns the index for the first match of a value in the given list or -1 if the value is not present.
```

> | *Description* 3+| Returns the index for the first match of a value in the given list or -1 if the value is not present.

Reviewer: ```suggestion
| *Description* 3+| Returns the index of the first match of a value in the given list or -1 if the value is not present.
```

.3+| *Arguments* | *Name* | *Type* | *Description*
| `list` | `LIST<ANY>` | A list to be searched.
| `value` | `ANY` | A value to search for.
| *Returns* 3+| `INTEGER`
|===

.Considerations
|===

| Indexing is 0-based.
| Returns -1 if the value is not present in the list.
| `coll.indexOf(null, null)` returns `null`.
| `coll.indexOf(null, value)` returns `null`.
| `coll.indexOf(list, null)` returns `null`.

|===

.+coll.indexOf()+ 
======

.Find the first index of a given value
// tag::functions_list_indexOf[]
[source, cypher]
----
RETURN coll.indexOf(['a', 'b', 'c', 'c'], 'c')
----
// end::functions_list_indexOf[]

An `INTEGER` representing the index of the item in the list.

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| coll.indexOf(['a', 'b', 'c', 'c'], 'c')
| 2
1+d|Rows: 1
|===

.Search for a value not present in the given list

> // tag::functions_list_indexOf_missing[]

Reviewer: ```suggestion
```

[source, cypher]
----
RETURN coll.indexOf([1, 'b', false], 4.3)
----

> // end::functions_list_indexOf_missing[]

Reviewer: ```suggestion
```

The `INTEGER` -1 representing that the value is not found.

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| coll.indexOf([1, 'b', false], 4.3)
| -1
1+d|Rows: 1
|===
======

[[functions-coll-insert]]
== coll.insert()

.Details
|===
| *Syntax* 3+| `coll.insert(list, index, value)`
| *Description* 3+| Returns a list with the given value inserted at the given index.
.4+| *Arguments* | *Name* | *Type* | *Description*
| `list` | `LIST<ANY>` | A list to add to.
| `index` | `INTEGER` | The index to add the given value at.
| `value` | `ANY` | The value to add into the list.
| *Returns* 3+| `LIST<ANY>`
|===

.Considerations
|===
| Indexing is 0-based.
| `coll.insert(null, null, value)` returns `null`.
| `coll.insert(list, null, value)` returns `null`.
| `coll.insert(null, 1, value)` returns `null`.

> | If the given `index` is negative or larger than the size of the given list, an error will be returned.

Reviewer: ```suggestion
| If the given `index` is negative or larger than the size of the given list, an error is returned.
```

Reviewer: is the error returned or thrown?

|===

.+coll.insert()+
======

.Query
// tag::functions_list_insert[]
[source, cypher]
----
RETURN coll.insert([true, 'a', 1, 5.4], 1, false)
----
// end::functions_list_insert[]

> The original list with a new value inserted at the given index. 

Reviewer: since coll.remove talks about shifting, is that aspect relevant here as well?

Author: Yes, it probably is, how about:

The original list with a new value inserted at the given index, shifting all the values following by 1.

| *Returns* 3+| `ANY`
|===

.Considerations
|===
| `coll.min(null)` returns `null`.
| `coll.min([])` returns `null`.
| The minimum is determined using Cypher's standard value ordering; see xref:values-and-types/ordering-equality-comparison.adoc[].
|===

.+coll.min()+ 
======

.Query
// tag::functions_list_min[]
[source, cypher]
----
RETURN coll.min([true, 'a', 1, 5.4])
----
// end::functions_list_min[]

The minimum value found in the given list based on Cypher's ordering.

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===
| coll.min([true, 'a', 1, 5.4])
| true
1+d|Rows: 1
|===

======

[[functions-coll-remove]]
== coll.remove()

.Details
|===
| *Syntax* 3+| `coll.remove(list, index)`
| *Description* 3+| Returns a list with the value at the given index removed.
.3+| *Arguments* | *Name* | *Type* | *Description*
| `list` | `LIST<ANY>` | A list to remove a value from.
| `index` | `INTEGER` | The index of the value to be removed.
| *Returns* 3+| `LIST<ANY>`
|===

.Considerations
|===
| Indexing is 0-based.
| `coll.remove(null, null)` returns `null`.
| `coll.remove(null, index)` returns `null`.
| `coll.remove(list, null)` returns `null`.

> | If the given `index` is negative or larger than the size of the given list, an error will be returned.

Reviewer: ```suggestion
| If the given `index` is negative or larger than the size of the given list, an error is returned.
```

|===

.+coll.remove()+
======

.Query
// tag::functions_list_remove[]
[source, cypher]
----
RETURN coll.remove([true, 'a', 1, 5.4], 1)
----
// end::functions_list_remove[]

> The original list with the value at the given index removed and all items following shifted forward by 1.

Reviewer: ```suggestion
The original list with the value at the given index removed and all items following shifted by 1.
```
unsure about "forward", but maybe we can just dodge it entirely?

> The original list with the value at the given index removed and all items following shifted by 1.

Reviewer: ```suggestion
The original list with the value at the given index removed and all values following shifted by 1.
```
(for consistency)
